### PR TITLE
[Core]Async updating issue fixed for actor's num_restart

### PR DESCRIPTION
**Pull request by lixin-wei**
## Why are these changes needed?

num_restarts must be set before state, or the RESTARTING actor notification might contain a wrong num_restarts. Because GCS is updated asynchronously.

This will cause a critical problem that I met today:

When an actor restarts, since the problem mentioned, RESTARTING notification might contain a lagger num_restarts than expected.

For example, when an actor died and restarts for the first time,  all other actors might receive a RESTARTING notification containing num_restarts=1(which is expected to be 0).

Then, in `direct_actor_transport.cc  CoreWorkerDirectActorTaskSubmitter::DisconnectActor`, `num_restars` in ClinetQueue will be set to 1.  

Then ALIVE notification arrives, but in `direct_actor_transport.cc  CoreWorkerDirectActorTaskSubmitter::ConnectActor`, it will be skipped since `1<=1`
```
  if (num_restarts <= queue->second.num_restarts) {
    // This message is about an old version of the actor and the actor has
    // already restarted since then. Skip the connection.
    return;
  }
```

At last, the restarted actor can't be called and won't receive any task, since no rpc_client is available.

## My fix approach

The easiest way I think is to set num_restars before updating the state to RESTARTING.

After that RESTARTING notification will always contain num_restarts that are already updated.

Then in the handler of RESTARTING notification, num_restart should be minus by 1.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

**Comment by lixin-wei** [comment]
@stephanie-wang @kfstorm @raulchen Could you please review this PR?

**Comment by lixin-wei** [comment]
@stephanie-wang Hi, could you please review this PR, this blocks some new features of ours.

File: src/ray/gcs/gcs_server/gcs_actor_manager.cc [hunk 1]
```diff
@@ -793,6 +793,9 @@ void GcsActorManager::ReconstructActor(const ActorID &actor_id, bool need_resche
                    << " at node " << node_id << ", need_reschedule = " << need_reschedule
                    << ", remaining_restarts = " << remaining_restarts;
   if (remaining_restarts != 0) {
+    // num_restarts must be set before updating GSC, or the RESTARTING
```
Comment (stephanie-wang) [inline src/ray/gcs/gcs_server/gcs_actor_manager.cc:796]:
```suggestion
    // num_restarts must be set before updating GCS, or the RESTARTING
```

**Comment by stephanie-wang** [comment]
Sorry for the delay.

I actually don't understand how the problem that you described can come up. Specifically, how is it possible for this part to happen?
> For example, when an actor died and restarts for the first time, all other actors might receive a RESTARTING notification containing num_restarts=1(which is expected to be 0).

According to the code, it seems like `num_restarts` doesn't get incremented until *after* the GCS publishes the RESTARTING message.

**Comment by lixin-wei** [comment]
> Sorry for the delay.
> 
> I actually don't understand how the problem that you described can come up. Specifically, how is it possible for this part to happen?
> 
> > For example, when an actor died and restarts for the first time, all other actors might receive a RESTARTING notification containing num_restarts=1(which is expected to be 0).
> 
> According to the code, it seems like `num_restarts` doesn't get incremented until _after_ the GCS publishes the RESTARTING message. So shouldn't the other actors receive a RESTARTING notification with `num_restarts=0`?

Because `gcs_table_storage_->ActorTable().Put()` is asynchronous.

**Comment by stephanie-wang** [comment]
I still don't understand. Doesn't this line copy the non-incremented value for num_restarts? https://github.com/ray-project/ray/blob/master/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L804

If not, I think the right solution is to increment the value once we're sure that the value has been put.

**Comment by lixin-wei** [comment]
@stephanie-wang Yes, hold on, although `gcs_table_storage_->ActorTable().Put()` is asynchronous, the published data in it is a copy, and num_restarts should be 0. There should be something that went wrong with pub-sub. Give me some time to dig it out.

**Comment by lixin-wei** [comment]
@stephanie-wang Finally my colleague found it. On the subscription side, it will get GCS's data at the first time. The published data actually not be received by the subscription side, instead, it will get the asynchronously written data. That's why it will receive a notification contains num_restarts=1.

So, the solution seems to be the same as the current: update num_restarts before publishing it.

BTW, why did you update num_restarts after publishing, did you have other consideration?

**Comment by lixin-wei** [comment]
@stephanie-wang So, the right solution seems to update `num_restarts` in ALIVE publishing but not in RESTARTING. What do you think?

By the way, why did you put num_restarts after publishing? Did you have other intentions or it's a mistake?

**Comment by stephanie-wang** [comment]
Hmm sorry I still don't really understand. Can you link to the code that reads the data asynchronously?

> @stephanie-wang So, the right solution seems to update num_restarts in ALIVE publishing but not in RESTARTING. What do you think?

I'm not necessarily opposed to it, but I would like to understand the problem first. From the code, it doesn't look like the scenario is even possible, so that must mean I'm misunderstanding something about how pubsub works. If that's true, then we should document this because it's very unintuitive that this could happen.

>
> By the way, why did you put num_restarts after publishing? Did you have other intentions or it's a mistake? Will this solution break any other thing?

It's just to make sure the idempotence handling on the client side is correct. So it should always be:
```
ALIVE:0 RESTARTING:0 ALIVE:1 RESTARTING:1, etc
```

But it also seems fine to do
```
ALIVE:0 RESTARTING:1 ALIVE:1 RESTARTING:2, etc
```

If we do end up doing this, please make sure to document it and update the unit test cases for direct_actor_transport.cc.

**Comment by kfstorm** [comment]
@stephanie-wang The thing is, after `num_restarts` was updated in local cache here https://github.com/ray-project/ray/blob/946ae74817835fb82f86ae2fdf4f12a19f5c7885/src/ray/gcs/gcs_server/gcs_actor_manager.cc#L811, the cached data may be flushed to external storage in https://github.com/ray-project/ray/blob/36e626e95d7cb905708bb997fc102ad1d91eb0be/src/ray/gcs/gcs_server/gcs_actor_scheduler.cc#L311. Later when a client subscribe to this actor, it will get the data from external storage, which is 1 here.

So we'd better avoid the inconsistency between local cache and external storage.

**Comment by lixin-wei** [comment]
@stephanie-wang The read code is here

https://github.com/lixin-wei/ray/blob/num_restart_fix/src/ray/gcs/gcs_client/service_based_accessor.cc#L316

On subscription, the client will first get GCS's data to generate a fake "notification". The data is not the published one but is what actually in storage.

**Comment by lixin-wei** [comment]
@kfstorm @stephanie-wang So there are two solutions here:

1. Updating num_restarts before publishing it in RESTARTING. then it will be: `ALIVE:0 RESTARTING:1 ALIVE:1 RESTARTING:2` And num_restarts in the handler of RESTARTING should be minus by 1. The code change will be the same as current.

2. Move num_restarts updating to ALIVE and change the default value of num_restars to -1. This will cause a massive scale change of codes.  All through table creating, task handling...

3. Move num_restarts updating to ALIVE, add a field to actor table to determine whether it is the first time the actor is ALIVE. If not, plus the num_restarts by 1.

I prefer the first solution, what do you think?

**Comment by lixin-wei** [comment]
@stephanie-wang Hi, could you please review again?

**Comment by stephanie-wang** [comment]
I see, thanks, that makes sense!

Instead of the logic to decrement the num_starts field, can you just swap the < and <= checks [here](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/transport/direct_actor_transport.cc#L131) and [here](https://github.com/ray-project/ray/blob/master/src/ray/core_worker/transport/direct_actor_transport.cc#L171)? That should have the same effect, but I think it's a bit easier to understand.

Can you also update the documentation to reflect this? I would add comments to the code linked above and also document the expected num_restarts values in `gcs.proto`.

Also, can you update the unit tests? I think the main relevant one is this one: https://github.com/ray-project/ray/blob/master/src/ray/core_worker/test/direct_actor_transport_test.cc#L338

**Comment by lixin-wei** [comment]
@stephanie-wang All done. You already had comments in `direct_actor_transport.cc`, so I only add comments in gcs.proto. Please have a look.

**Review by stephanie-wang — APPROVED** [review]
Looks great, thanks for finding this issue!

You can remove the author-action-required tag once tests are passing, and I'll merge.

**Comment by lixin-wei** [comment]
The failed test seems not related to my modification, @stephanie-wang could you please double-check it?

**Pull request merged**